Remotion/Web/ClientScript/Scripts/DropDownMenu.ts

```diff
@@ -0,0 +1,928 @@
+  private static _itemClicked: boolean = false;
+  private static _repositionInterval: number = 200;
+  private static _repositionTimer: Nullable<number> = null;
+  private static _statusPopupDelay: number = 200;
+  private static _statusPopupRepositionTimer: Nullable<number> = null;
+  private static _blurTimer: Nullable<number> = null;
+  private static _updateFocus: boolean = true;
+
+  public static AddMenuInfo(menuInfo: DropDownMenu_MenuInfo): void
+  {
+    DropDownMenu._menuInfos[menuInfo.ID] = menuInfo;
+  }
+
+  public static BindOpenEvent (openTarget: JQuery, menuID: string, eventType: string, getSelectionCount: Nullable<() => number>, moveToMousePosition: boolean): void
+  {
+    ArgumentUtility.CheckNotNullAndTypeIsJQuery ('openTarget', openTarget);
+    ArgumentUtility.CheckNotNullAndTypeIsString ('menuID', menuID);
+    ArgumentUtility.CheckNotNullAndTypeIsString ('eventType', eventType);
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean ('moveToMousePosition', moveToMousePosition);
+
+    var menuButton = openTarget[0].tagName.toLowerCase() === 'a' ? openTarget : openTarget.find ('a[href]:last');
+    menuButton.attr('aria-haspopup', 'menu');
+    menuButton.attr('aria-expanded', 'false');
+
+    openTarget.bind(
+        eventType,
+        function (evt: JQueryMouseEventObject)
+        {
+          menuButton.focus();
+          DropDownMenu.OnClick (openTarget, menuID, getSelectionCount, moveToMousePosition ? evt : null);
+        });
+
+    if (!moveToMousePosition)
+    {
+      menuButton
+          .bind ("focus", function (evt) { openTarget.find ('.' + DropDownMenu._buttonClassName).addClass (DropDownMenu._focusClassName); })
+          .bind("blur", function (evt) { openTarget.find('.' + DropDownMenu._buttonClassName).removeClass (DropDownMenu._focusClassName); });
+
+      var allButMenuButton = openTarget.find ('a[href]').not(menuButton);
+      allButMenuButton
+          .bind ("mouseover", function (evt) { openTarget.find ('.' + DropDownMenu._buttonClassName).addClass (DropDownMenu._nestedHoverClassName); })
+          .bind ("mouseout", function (evt) { openTarget.find ('.' + DropDownMenu._buttonClassName).removeClass (DropDownMenu._nestedHoverClassName); });
+    }
+  }
+
+  public static LoadFilteredMenuItems(
+      itemInfos: DropDownMenu_ItemInfo[],
+      loadMenuItemStatus: Nullable<(s: (i: DropDownMenu_ItemInfo[]) => void, e: () => void) => void>,
+      onSuccess: (i: DropDownMenu_ItemInfo[]) => void,
+      onError: () => void): void
+  {
+    if (loadMenuItemStatus !== null)
+    {
+      const currentLoadOperationID = new Object();
+      const onSuccessForLoadMenuItemStatus = function (itemStatusArray: DropDownMenu_ItemInfo[])
+      {
+        if (DropDownMenu._currentLoadOperation !== currentLoadOperationID)
+          return;
+        DropDownMenu._currentLoadOperation = null;
+
+        const itemStatusMap = itemStatusArray.reduce(
+          function (map: Dictionary<any>, itemStatus)
```

Comment:
Should be `Dictionary<DropDownMenu_ItemInfo>` and we should annotate the second parameter as well

Reply:
Whoops, why is there no transpiler error (itemStatus should be annotated) like for normal methods?

Reply:
Because a the supplied function can be converted to the implied function. `Dictionary<any>` can be reduced to `Dictionary<BodyList_InfoItem>`

Reply:
I meant the missing annotation on the second parameter

Reply:
`reduce` can infer the parameter type from `itemStatusArray`.

```ts
interface Array<T> {
  reduce<U>(callbackfn: (previousValue: U, currentValue: T, currentIndex: number, array: T[]) => U, initialValue: U): U;
}
```

with `U` being the type we are mapping to (`Dictionary<DropDownMenu_ItemInfo>`) and `T` being the type of the array.